### ISSUE TITLE
Predicting probabilities of each label outputted in arow.py

### ISSUE THREAD

**Issue by dhruvghulati-zz** [andreasvlachos/arow_csc#5]
I am trying to predict the probabilities the system has outputted in AROW for each label in my test instances.

I am using this type of test data (it is a naive baseline):

```
1:1 2:1 3:1 4:1 5:1 6:1 7:1 8:1 9:1 10:1 11:0 12:1 13:1 14:1 15:1 16:1 | location_slot turkey france netherlands number_slot lrb rrb
```

And using this code:

```
train_data = [arow.train_instance_from_svm_input(line) for line in trainDataLines]
test_data = [arow.test_instance_from_svm_input(line) for line in testDataLines]
cl = arow.AROW()
cl.train(train_data)
cl.probGeneration()
predictions = [cl.predict(d, verbose=True,probabilities=True).label2prob for d in test_data]
```

I am getting these probabilities (each line is roughly like this):

```
{'11': 1e-06, '10': 1e-06, '13': 1e-06, '12': 1e-06, '15': 1e-06, '14': 1e-06, '16': 1e-06, '1': 1e-06, '3': 1e-06, '2': 1e-06, '5': 1e-06, '4': 1e-06, '7': 1e-06, '6': 1e-06, '9': 1.0000010000000001, '8': 1e-06}
```

Which is clearly matching my 0 and 1 costs I have for my input data. 

Whereas when I use ordinary logistic regression in scikit-learn (where I don't have costs per label, just one label per instance), outputting `predict_proba`, I am getting:

```
5.085409071212559917e-02 1.120198203180571873e-02 3.857035939484883269e-02 2.084571186163139411e-02 1.297530078841703863e-02 1.878877233829619881e-02 8.501321511167143907e-02 2.107530829257814667e-02 4.661419888723658533e-02 3.110850980702436347e-02 3.278946527730437766e-02 8.223427089410631496e-02 3.984204282790664298e-01 8.036492304852346780e-02 2.710335198634480219e-02 4.204011128901925798e-02
```

Am I doing something wrong with `probGeneration`? Should I be using `batchPredict` or something to adapt the probabilities to be something more intelligent?

**Comment by andreasvlachos** [comment]
While probabilistic output is supported by the code (and theory of AROW), it is a bit tricky to get it right due an extra parameter that needs tuning. Most of the times what is needed is some sort of score per label, which would allow the use of a threshold say to adjust precision/recall. Score per label are produced in the predict function and stored in the label2score of the prediction objects.

**Comment by dhruvghulati-zz** [comment]
Brilliant! I am also presuming higher score is better just from briefly looking at the code? Is there a range the scores will generally be in so I can set up a hyperparameter range that makes sense?

Just taking the example above, the `label2scores` for each instance and the labels across the instance seem to be super similar, and cluster around a very different range for each row e.g (22 and 36 in the range below).:

```
{'11': 22.03245531058002, '10': 21.92074785118018, '13': 21.9154854003752, '12': 21.72169405207605, '15': 21.763820945091, '14': 21.872857638241914, '16': 18.03990187862184, '1': 21.85483414800496, '3': 21.85310968009222, '2': 21.94306120094102, '5': 21.62829739929052, '4': 21.910852915353303, '7': 21.81557832422554, '6': 22.015549641726906, '9': 21.958855764802767, '8': 20.57187757122006}
```

And then next row:

```
{'11': 36.883248120218404, '10': 36.732969712018566, '13': 36.7152050442673, '12': 36.53344118402834, '15': 36.46534601866472, '14': 36.75394255891222, '16': 33.832925032621986, '1': 36.60666949962299, '3': 36.77700981488911, '2': 36.735534884788294, '5': 36.335625929930394, '4': 36.77201641855946, '7': 36.723409208179035, '6': 36.85516383136266, '9': 36.798053173100165, '8': 35.580321644898596}
```

So does it make sense creating different thresholds per row (e.g. if the best one is >30% higher than the next best in the row, it is a label, else it is not counted as a claim label), or is it OK to make a universal threshold across rows e.g. "30" in which case row 2's best label would be the label, row 1 would be labelled as "no_property"?

**Comment by andreasvlachos** [comment]
Higher scores are better indeed. The scores can be anything, thus any threshold should be defined by using a dev set. I don't think it makes sense to define them per-row (how would you be able to know the threshold for an unseen instance?), but a universal threshold would be fine. Just remember that such a threshold would need to be retuned every time a model is trained.

**Comment by dhruvghulati-zz** [comment]
OK. One idea I had was to normalise each dict of the above by 0 and 1 by either min-max, or by a sigmoid function `(1/1+exp(-x)))` with no bias or constant term. That way you could optimise a threshold on a dev set by knowing the range in advance of all potential values (0 to 1). Do you agree this is a better way? And if so, what would be the better way to normalise - min-max, or sigmoid, or does it not matter?

**Comment by andreasvlachos** [comment]
I don't think normalizing will help in any way; the scores will just have a restricted range, but they still have no meaning. Being in the range of 0 to 1 does not mean that that 0.5 has max uncertainty. Not having a known range doesn't make the optimisation of the threshold on a dev set easier or harder.
